File: 0079-word-search/0079-word-search.cpp
class Solution {
public:
    bool solve(int i,int j,int l,int m,int n,string &word,vector<vector<char>>&board,vector<vector<int>>&vis){
        if(l>=word.size()){
            return true;
        }
        if(i<0 || j<0 || i>=m || j>=n || vis[i][j]){
            return false;
        }
        if(board[i][j]==word[l]){
            vis[i][j]=true;
            bool left=solve(i,j-1,l+1,m,n,word,board,vis);
            bool up = solve(i-1,j,l+1,m,n,word,board,vis);
            bool right = solve(i,j+1,l+1,m,n,word,board,vis);
            bool down =solve(i+1,j,l+1,m, n,word,board,vis);
            if(left || up || right || down){
                return true;
            }
        }
        vis[i][j]=false;
        return false;
    }
    bool exist(vector<vector<char>>& board, string word) {
        int m=board.size();
        int n=board[0].size();
        vector<vector<int>>vis(m,vector<int>(n,0));
        for(int i=0;i<m;i++){
            for(int j=0;j<n;j++){
                if(board[i][j]==word[0]){
                    if(solve(i,j,0,m,n,word,board,vis)){
                        return true;
                    }
                }
            }
        }
        return false;
    }
};

File: 0079-word-search/0079-word-search.java
class Solution {
    public boolean solve(int i,int j,int len,String word,char[][] board,int n,int m,boolean[][] vis){
        if(len>=word.length()){
            return true;
        }
        if(i<0 || j<0 || i>=n || j>=m || vis[i][j]==true){
            return false;
        }
        if(word.charAt(len)==board[i][j]){
            vis[i][j]=true;
             boolean left = solve(i,j-1,len+1,word,board,n,m,vis);
             boolean up = solve(i-1,j,len+1,word,board,n,m,vis);
             boolean right = solve(i,j+1,len+1,word,board,n,m,vis);
             boolean down = solve(i+1,j,len+1,word,board,n,m,vis);
            if(left || up || right || down){
                return true;
            }
        }
        vis[i][j]=false;
        return false;
    }
    public boolean exist(char[][] board, String word) {
        int n=board.length;
        int m=board[0].length;
        boolean[][] vis = new boolean[n][m];
        
        for(int i=0;i<n;i++)
            for(int j=0;j<m;j++)
                vis[i][j]=false;
        
        for(int i=0;i<n;i++){
            for(int j=0;j<m;j++){
                if(board[i][j]==word.charAt(0)){
                    if(solve(i,j,0,word,board,n,m,vis)){
                        return true;
                    }
                }
            }
        }
        return false;
    }
}